From 9f48286a07354a472028620b4de0a69182930e56 Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Mon, 4 Aug 2014 19:04:58 -0400
Subject: [fs] btrfs: fix race between balance recovery and root deletion

Message-id: <1407179109-13714-412-git-send-email-sandeen@redhat.com>
Patchwork-id: 87343
O-Subject: [RHEL7.1 PATCH 411/422] Btrfs: fix race between balance recovery and root deletion
Bugzilla: 1071179
RH-Acked-by: Zach Brown <zab@redhat.com>

From: Wang Shilong <wangsl.fnst@cn.fujitsu.com>

commit 5f3164813b90f7dbcb5c3ab9006906222ce471b7
Author: Wang Shilong <wangsl.fnst@cn.fujitsu.com>
Date:   Thu Jun 26 11:08:16 2014 +0800

    Btrfs: fix race between balance recovery and root deletion

    Balance recovery is called when RW mounting or remounting from
    RO to RW, it is called to finish roots merging.

    When doing balance recovery, relocation root's corresponding
    fs root(whose root refs is 0) might be destroyed by cleaner
    thread, this will make btrfs fail to mount.

    Fix this problem by holding @cleaner_mutex when doing balance
    recovery.

    Signed-off-by: Wang Shilong <wangsl.fnst@cn.fujitsu.com>
    Signed-off-by: Chris Mason <clm@fb.com>
Signed-off-by: Eric Sandeen <sandeen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/btrfs/disk-io.c b/fs/btrfs/disk-io.c
index dedecf8..7c748a0 100644
--- a/fs/btrfs/disk-io.c
+++ b/fs/btrfs/disk-io.c
@@ -2907,7 +2907,9 @@ retry_root_backup:
   if (ret)
    goto fail_qgroup;
 
+  mutex_lock(&fs_info->cleaner_mutex);
   ret = btrfs_recover_relocation(tree_root);
+  mutex_unlock(&fs_info->cleaner_mutex);
   if (ret < 0) {
    printk(KERN_WARNING
           "BTRFS: failed to recover relocation\n");
diff --git a/fs/btrfs/super.c b/fs/btrfs/super.c
index 35e08b2..e2f3b01 100644
--- a/fs/btrfs/super.c
+++ b/fs/btrfs/super.c
@@ -1466,7 +1466,9 @@ static int btrfs_remount(struct super_block *sb, int *flags, char *data)
    goto restore;
 
   /* recover relocation */
+  mutex_lock(&fs_info->cleaner_mutex);
   ret = btrfs_recover_relocation(root);
+  mutex_unlock(&fs_info->cleaner_mutex);
   if (ret)
    goto restore;
 
-- 
1.7.1